### PR TITLE
Enabled debugging services in Roslyn branch

### PR DESCRIPTION
**Pull request by otawfik-ms**
Added the following two services:

- BreakpointResolutionService: which locates the correct span a breakpoint should be placed at.
- LanguageDebugInfoService: which displays a data tip over variables when in debugging mode.

Screenshot below:

image

### CODE REVIEW

**Comment by otawfik-ms** [comment]
@Microsoft/fsharp-compiler

File: vsintegration/src/FSharp.Editor/LanguageDebugInfoService.fs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                let! sourceText = document.GetTextAsync(cancellationToken) |> Async.AwaitTask
+                let textSpan = TextSpan.FromBounds(0, sourceText.Length)
+                let tokens = FSharpColorizationService.GetColorizationData(sourceText, textSpan, Some(document.Name), defines, cancellationToken)
+
+                return match FSharpLanguageDebugInfoService.GetDataTipInformation(sourceText, position, tokens) with
+                       | None -> Unchecked.defaultof<DebugDataTipInfo>
+                       | Some(textSpan) -> new DebugDataTipInfo(textSpan, sourceText.GetSubText(textSpan).ToString())
+            }
+
+            Async.StartAsTask(computation, TaskCreationOptions.None, cancellationToken).ContinueWith(fun(task: Task<DebugDataTipInfo>) ->
```
Comment (OkayX6) [inline vsintegration/src/FSharp.Editor/LanguageDebugInfoService.fs]: Factor out this Async.StartAsTask() pattern maybe? I'm seeing it three times.

File: vsintegration/src/FSharp.Editor/BraceMatchingService.fs [hunk 2]
```diff
@@ -3,117 +3,42 @@
...
-                                    result <- Some(BraceMatchingResult(TextSpan(min position currentPosition, 1), TextSpan(max position currentPosition, 1)))
-                                else
-                                    braceDepth <- braceDepth - 1
-                        currentPosition <- proceedFunc currentPosition
-                    result
+    static member GetBraceMatchingResult(sourceText, fileName, options, position) = async {
+        let isPositionInRange(range) =
+            let span = CommonRoslynHelpers.FSharpRangeToTextSpan(sourceText, range)
+            span.Start <= position && position <= span.End
+        let! matchedBraces = FSharpChecker.Instance.MatchBracesAlternate(fileName, sourceText.ToString(), options)
```
Comment (jaredpar) [inline vsintegration/src/FSharp.Editor/BraceMatchingService.fs]: > FSharpChecker.Instance [](start = 28, length = 23)

Is FSharpChecker an immutable type?  If not and it contains mutable state consider passing this as a parameter.

File: vsintegration/src/FSharp.Editor/BreakpointResolutionService.fs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+open Microsoft.FSharp.Compiler.Parser
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open Microsoft.FSharp.Compiler.Range
+
+[<Shared>]
+[<ExportLanguageService(typeof<IBreakpointResolutionService>, FSharpCommonConstants.FSharpLanguageName)>]
+type internal FSharpBreakpointResolutionService() =
+
+    static member GetBreakpointLocation(sourceText: SourceText, fileName: string, textSpan: TextSpan, options: FSharpProjectOptions) = async {
+        let! parseResults = FSharpChecker.Instance.ParseFileInProject(fileName, sourceText.ToString(), options)
```
Comment (jaredpar) [inline vsintegration/src/FSharp.Editor/BreakpointResolutionService.fs]: >  let! parseResults = FSharpChecker.Instance.ParseFileInProject(fileName, sourceText.ToString(), options) [](start = 6, length = 105)

This is going to cause a large allocation every time due to the ToString.  Are the results of parsing cached anywhere based on a SourceText value?  For example if parsing a file with the same checksum just return previous parse.

File: vsintegration/src/FSharp.Editor/BreakpointResolutionService.fs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                       | Some(range) -> BreakpointResolutionResult.CreateSpanResult(document, CommonRoslynHelpers.FSharpRangeToTextSpan(sourceText, range))
+            }
+
+            Async.StartAsTask(computation, TaskCreationOptions.None, cancellationToken).ContinueWith(fun(task: Task<BreakpointResolutionResult>) ->
+                if task.Status = TaskStatus.RanToCompletion then
+                    task.Result
+                else
+                    Assert.Exception(task.Exception.GetBaseException())
+                    raise(task.Exception.GetBaseException())
+            , cancellationToken)
```
Comment (jaredpar) [inline vsintegration/src/FSharp.Editor/BreakpointResolutionService.fs]: This helper seems to be used a lot.  Factor it out into a separate function?

File: vsintegration/src/FSharp.Editor/BraceMatchingService.fs (hunk 2 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.Editor/BraceMatchingService.fs]: Static helper methods in service.fsi for parsing/type checking.

File: vsintegration/src/FSharp.Editor/BreakpointResolutionService.fs (hunk 3 above)
Comment (otawfik-ms) [inline vsintegration/src/FSharp.Editor/BreakpointResolutionService.fs]: Good point. I don't see it being done on SourceText, just a memory allocation. I'll need to look into caching on our side if there was a performance issue in UX.

**Comment by CyrusNajmabadi** [comment]
Looks ok to me.

**Pull request merged**